### ISSUE TITLE
Action loss is not declining on LISA training

### ISSUE THREAD

**Issue by dljzx** [Div99/LISA#1]
Thanks for your work! While i was training LISA model on lorel dataset with 50000 episodes, I found out that the action loss did not efficiently decline while commitment loss declines successfully . So that the total loss remaining about 0.93 did not decrease after 30 steps. So It makes evaluation success keep on zero after 2500 iterations. Can you help me?

**Comment by dljzx** [comment]
After training with LISA, I tried using lang DT to train with the same dataset. Although action pred loss successfully declined, the success rate still remaining below 0.2 and no upward trend through 2500 iterations. Is that error on my dataset or the code? 

![image](https://user-images.githubusercontent.com/47135039/220072257-42b4218a-fd0e-42f9-ada9-aa3137d2d51b.png)
success rate by lang DT

![image](https://user-images.githubusercontent.com/47135039/220072494-f01ce958-b16c-4233-a43f-2c33b3a0b608.png)
success rate by lisa

main.py model=traj_option batch_size=1024 option_selector.use_vq=True seed=0 train_dataset.num_trajectories=50000 model.horizon=10 model.K=10 option_selector.num_options=20 env=lorel_sawyer_state warmup_steps=2500 max_iters=2500 trainer.eval_every=100 option_selector.commitment_weight=0.25 option_selector.kmeans_init=True save_interval=100
and here are my hyperparameters

**Comment by skandavaidyanath** [comment]
Hi, thanks so much for your interest in our work and apologies that you're having trouble with reproducing the results. A couple of pointers that might help you debug this issue:
1. The success rates in Table 1 of the paper are for the 6 original Lorl instructions without any kind of rephrasing. The evaluation done by the code is over all kinds of rephrasing so this number will be lower than reported in the paper. There are other scripts like `viz-lorl.py` and `viz-lorl.ipynb` that evaluate on only the 6 main instructions or you can easily modify the code to get this.
2. When we used the Lorl data, we cleaned it by removing all the "Do nothing" labeled instructions. This resulted in us having only 40108 trajectories and not 50k. I'm not sure if this should lead to such a big disparity in results though.
3. Getting a success rate of 0 on this environment is weird because even a random policy should get some success on this tabletop environment. Maybe there's a bug you've introduced somehow?

Here are some wandb plots attached for Lorl with states to get an indication of what the curves should look like:

**For Lang DT:**
![Screenshot 2023-02-20 at 3 52 18 PM](https://user-images.githubusercontent.com/34721126/220214508-394e4bda-9297-4d36-91e1-22cc92242025.png)
![Screenshot 2023-02-20 at 3 52 29 PM](https://user-images.githubusercontent.com/34721126/220214521-3e4492cb-d422-4b0b-8f94-89455c0fe69d.png)

**For LISA:**
![Screenshot 2023-02-20 at 3 53 32 PM](https://user-images.githubusercontent.com/34721126/220214610-31c58c90-7664-4d82-a88a-4d34bd8926d0.png)
![Screenshot 2023-02-20 at 3 53 40 PM](https://user-images.githubusercontent.com/34721126/220214625-6ce51af0-c6c8-495f-a370-e96ce9515fbb.png)

From the curves and your numbers, it seems like the Lang DT success rates are similar and the loss numbers are similar. Maybe there's a bug in the evaluation script for LISA that's crept into the code?

Please let us know if this helps and if there's anything else we can do to resolve this issue!

**Comment by dljzx** [comment]
> Hi, thanks so much for your interest in our work and apologies that you're having trouble with reproducing the results. A couple of pointers that might help you debug this issue:
> 
> 1. The success rates in Table 1 of the paper are for the 6 original Lorl instructions without any kind of rephrasing. The evaluation done by the code is over all kinds of rephrasing so this number will be lower than reported in the paper. There are other scripts like `viz-lorl.py` and `viz-lorl.ipynb` that evaluate on only the 6 main instructions or you can easily modify the code to get this.
> 2. When we used the Lorl data, we cleaned it by removing all the "Do nothing" labeled instructions. This resulted in us having only 40108 trajectories and not 50k. I'm not sure if this should lead to such a big disparity in results though.
> 3. Getting a success rate of 0 on this environment is weird because even a random policy should get some success on this tabletop environment. Maybe there's a bug you've introduced somehow?
> 
> Here are some wandb plots attached for Lorl with states to get an indication of what the curves should look like:
> 
> **For Lang DT:** ![Screenshot 2023-02-20 at 3 52 18 PM](https://user-images.githubusercontent.com/34721126/220214508-394e4bda-9297-4d36-91e1-22cc92242025.png) ![Screenshot 2023-02-20 at 3 52 29 PM](https://user-images.githubusercontent.com/34721126/220214521-3e4492cb-d422-4b0b-8f94-89455c0fe69d.png)
> 
> **For LISA:** ![Screenshot 2023-02-20 at 3 53 32 PM](https://user-images.githubusercontent.com/34721126/220214610-31c58c90-7664-4d82-a88a-4d34bd8926d0.png) ![Screenshot 2023-02-20 at 3 53 40 PM](https://user-images.githubusercontent.com/34721126/220214625-6ce51af0-c6c8-495f-a370-e96ce9515fbb.png)
> 
> From the curves and your numbers, it seems like the Lang DT success rates are similar and the loss numbers are similar. Maybe there's a bug in the evaluation script for LISA that's crept into the code?
> 
> Please let us know if this helps and if there's anything else we can do to resolve this issue!

Thanks for your reply. Train curves of Lisa are similar so bugs are possibly on evaluation scipy.  The only different between Lisa and lang DT is in the get_action function. Can you recheck code of get_action and get_options?

**Comment by dljzx** [comment]
Additionally, I found that Word Freq matix output by my lisa training is terrible, through Train it fixed onto skillcode 11 for every task
![image](https://user-images.githubusercontent.com/47135039/220611807-b2f4f6d9-6401-427e-8327-d323bece4e50.png)

While I think that maybe VQVAE were not sucessfully trained. Maybe check the Option selector training would fix this problem

**Comment by skandavaidyanath** [comment]
Ah this picture helps a lot thanks! I think the code looks fine. Can you try a commitment weight of 1 for LoRL please?

**Comment by dljzx** [comment]
> 

It seems that commitment weight of 1 help a bit. Success rate no more zero but keep declining while training. On step 3 it achieves at about 0.3 success rate while with steps going more the success rate remain under 0.1. Also Word Freq matrix show only 8 and 18 two skills. 
![image](https://user-images.githubusercontent.com/47135039/221200445-74b69a83-5914-4912-8c2b-d8586b5a7ee4.png)
It is the image on step 250, Would that be better after more steps?

Also here is my command which I think might not suits your advice?
main.py env=lorel_sawyer_obs method=traj_option dt.n_layer=1 dt.n_head=4 option_selector.option_transformer.n_layer=1 option_selector.option_transformer.n_head=4 option_selector.commitment_weight=1.0 
option_selector.option_transformer.hidden_size=128 batch_size=1024 seed=2 warmup_steps=5000 trainer.eval_every=50 save_interval=50 max_iters=2500

**Comment by skandavaidyanath** [comment]
Yeah we ran it for 500 iterations but our options were more spread out by this time. Maybe run it for longer and see how it goes? I think your command looks fine to me. You should get an overall success rate of about 40% at the end. 
Also just to clarify, in my earlier images, those were for lorl_state since I thought you were running those. It seems like you're running lorl_observation now.

**Comment by dljzx** [comment]
> Yeah we ran it for 500 iterations but our options were more spread out by this time. Maybe run it for longer and see how it goes? I think your command looks fine to me. You should get an overall success rate of about 40% at the end. Also just to clarify, in my earlier images, those were for lorl_state since I thought you were running those. It seems like you're running lorl_observation now.

obs and state are both training on my mechine.While lorel_obs trains much slower than state so that there is no much result on lorel_obs for now. Lorel state receive bad results with low success rate and word freq matrix
![image](https://user-images.githubusercontent.com/47135039/221343673-b1062237-3987-406b-9a11-102cd858dd3a.png)
![image](https://user-images.githubusercontent.com/47135039/221343688-751cbcd2-2441-4043-9dc2-d7060bd9c662.png)

And here are mine command:
main.py env=lorel_sawyer_state method=traj_option dt.n_layer=1 dt.n_head=4 option_selector.option_transformer.n_layer=1 option_selector.option_transformer.n_head=4 option_selector.commitment_weight=1.0 option_selector.option_transformer.hidden_size=128 batch_size=256 seed=19 warmup_steps=2500 max_iters=2500

**Comment by dljzx** [comment]
If there was not bug  in the code , maybe I think there is difference between the lorel sawyer dataset. Can you share your dataset ?
